### PR TITLE
Fix bulk update with bool field

### LINKED ISSUE #239: Can't bulk update documents with boolean field type
```
> sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.14.1

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] dplyr_0.7.6       elastic_0.8.4     rlist_0.4.6.1     bindrcpp_0.2.2   
 [5] doMC_1.3.5        iterators_1.0.10  foreach_1.4.4     glue_1.3.0       
 [9] stringr_1.3.1     hdf5r_1.0.1       tidyr_0.8.1       aws.s3_0.3.12    
[13] readr_1.1.1       RPostgreSQL_0.6-2 DBI_0.8           dbplyr_1.2.1     
```

I'm using elasticsearch 6.2.3.

This doesn't work:

```r
library(elastic)
library(tibble)

d <- tibble(id = 1:3, x = c(TRUE, FALSE, TRUE), y = c("a", "b", "c"))

d_index <- '
{
  "mappings": {
    "d": {
      "properties": {
        "x": { "type": "boolean" },
        "y": { "type": "keyword" }
      }
    }
  }
}
'

if (index_exists("d")) index_delete("d")
index_create(index = "d", body = d_index)

res <- docs_bulk_update(
  d %>% select(-id),
  index = "d",
  type = "d",
  es_ids = F,
  doc_ids = d$id
)

print(res)
```

Output:

```
[[1]]
[[1]]$took
[1] 1

[[1]]$errors
[1] TRUE

[[1]]$items
[[1]]$items[[1]]
[[1]]$items[[1]]$update
[[1]]$items[[1]]$update$`_index`
[1] "d"

[[1]]$items[[1]]$update$`_type`
[1] "d"

[[1]]$items[[1]]$update$`_id`
[1] "1"

[[1]]$items[[1]]$update$status
[1] 400

[[1]]$items[[1]]$update$error
[[1]]$items[[1]]$update$error$type
[1] "mapper_parsing_exception"

[[1]]$items[[1]]$update$error$reason
[1] "failed to parse [x]"

[[1]]$items[[1]]$update$error$caused_by
[[1]]$items[[1]]$update$error$caused_by$type
[1] "illegal_argument_exception"

[[1]]$items[[1]]$update$error$caused_by$reason
[1] "Failed to parse value [ TRUE] as only [true] or [false] are allowed."

...
```

When initially uploading I can use `docs_bulk` instead of `docs_bulk_update` and this works. However, my use case requires updating.

What's really strange is that if I remove the keyword field so that my only field is the boolean one `x`, `docs_bulk_update` works again:

```r
d2_index <- '
{
  "mappings": {
    "d2": {
      "properties": {
        "x": { "type": "boolean" }
      }
    }
  }
}
'

if (index_exists("d2")) index_delete("d2")
index_create(index = "d2", body = d2_index)

d2 <- tibble(id = 1:3, x = c(TRUE, FALSE, TRUE))

res <- docs_bulk_update(
  d2 %>% select(-id),
  index = "d2",
  type = "d2",
  es_ids = F,
  doc_ids = d2$id
)

print(res)
```

Output:

```
[[1]]
[[1]]$took
[1] 5

[[1]]$errors
[1] FALSE

[[1]]$items
[[1]]$items[[1]]
[[1]]$items[[1]]$update
[[1]]$items[[1]]$update$`_index`
[1] "d2"

[[1]]$items[[1]]$update$`_type`
[1] "d2"

[[1]]$items[[1]]$update$`_id`
[1] "1"

[[1]]$items[[1]]$update$`_version`
[1] 1

[[1]]$items[[1]]$update$result
[1] "created"

[[1]]$items[[1]]$update$`_shards`
[[1]]$items[[1]]$update$`_shards`$total
[1] 2

[[1]]$items[[1]]$update$`_shards`$successful
[1] 2

[[1]]$items[[1]]$update$`_shards`$failed
[1] 0


[[1]]$items[[1]]$update$`_seq_no`
[1] 0

[[1]]$items[[1]]$update$`_primary_term`
[1] 1

[[1]]$items[[1]]$update$status
[1] 201

...
```

The only workaround I can come up with is to store "true" and "false" as strings, but this would require updating a lot of application code.

### PR DESCRIPTION
**Pull request by dpmccabe**
Fixes #239: In `docs_bulk_update` boolean fields get cast as strings. Just need to run `as.list` on each data.frame row, not on the entire data.frame.

### CODE REVIEW

File: R/docs_bulk_update.R [hunk 1]
```diff
@@ -111,10 +111,10 @@ make_bulk_update <- function(df, index, type, counter, path = NULL) {
     counter
   )
   
-  tmp <- apply(df, 1, as.list)
-  tmp <- lapply(unname(tmp), function(z) {
-    z$id <- NULL
-    list(doc = z, doc_as_upsert = TRUE)
+  tmp <- lapply(1:nrow(df), function(i) {
```
[the remark this answers is not in the document]
Comment (dpmccabe) [reply R/docs_bulk_update.R]: Always forget about that. Changed.

**Comment by dpmccabe** [comment]
I think there might be more to this fix in that the new code doesn't work with nested data. Looking into it...

File: R/docs_bulk_update.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (sckott) [reply R/docs_bulk_update.R]: thanks

**Comment by sckott** [comment]
you mean a list inside a data.frame?

**Comment by dpmccabe** [comment]
Yes, running `as.list` on each row messes up list columns. `make_bulk` doesn't have any issues here because it just calls 'toJSON` on the data.frame.

**Comment by sckott** [comment]
right, the update fxn is a little different as we need to set the id to null and set the doc as upsert field

**Comment by sckott** [comment]
if no test, can you at least show an example here with nested data

**Comment by dpmccabe** [comment]
My latest (hopefully final) commit should fix this. I think you actually need `dplyr` as a dependency if want want to test this. Here's some example data

```
> df
# A tibble: 3 x 3
  id                                       title                event_instances 
  <chr>                                    <chr>                <list>          
1 f9dd4b9105a8076c23d998897906e95f8dad33ee CCB Safety Training  <tibble [1 × 4]>
2 6b54d6f20c1b4c047658f4b3c78daf6dca81b6a6 Frankenreads         <tibble [1 × 4]>
3 42df3e4c8e743e927531f31d9ecd203e5bdb505e Nine Moments for Now <tibble [4 × 4]>

> df$event_instances
[[1]]
# A tibble: 1 x 4
  all_day dt                        end_dt location_text
  <lgl>   <chr>                     <chr>  <chr>        
1 NA      2017-12-20 14:00:00 -0400 NA     NA           

[[2]]
# A tibble: 1 x 4
  all_day dt                        end_dt                    location_text                                                                  
  <lgl>   <chr>                     <chr>                     <chr>                                                                          
1 FALSE   2018-10-31 09:00:00 -0400 2018-10-31 16:00:00 -0400 "Houghton Library, Edison and Newman Room\nQuincy St. & Harvard St., Cambridge"

[[3]]
# A tibble: 4 x 4
  all_day dt                        end_dt                    location_text                                                                                      
  <lgl>   <chr>                     <chr>                     <chr>                                                                                              
1 FALSE   2018-10-31 10:00:00 -0400 2018-10-31 17:00:00 -0400 "The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"
2 FALSE   2018-11-01 10:00:00 -0400 2018-11-01 17:00:00 -0400 "The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"
3 FALSE   2018-11-02 10:00:00 -0400 2018-11-02 17:00:00 -0400 "The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"
4 FALSE   2018-11-03 10:00:00 -0400 2018-11-03 17:00:00 -0400 "The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"
```

The new `make_bulk_update` function generates this:

```json
{"update":{"_index":"events","_type":"events","_id":"f9dd4b9105a8076c23d998897906e95f8dad33ee"}}
{"doc":{"title":"CCB Safety Training","event_instances":[{"all_day":null,"dt":"2017-12-20 14:00:00 -0400","end_dt":null,"location_text":null}]},"doc_as_upsert":true}
{"update":{"_index":"events","_type":"events","_id":"6b54d6f20c1b4c047658f4b3c78daf6dca81b6a6"}}
{"doc":{"title":"Frankenreads","event_instances":[{"all_day":false,"dt":"2018-10-31 09:00:00 -0400","end_dt":"2018-10-31 16:00:00 -0400","location_text":"Houghton Library, Edison and Newman Room\nQuincy St. & Harvard St., Cambridge"}]},"doc_as_upsert":true}
{"update":{"_index":"events","_type":"events","_id":"42df3e4c8e743e927531f31d9ecd203e5bdb505e"}}
{"doc":{"title":"Nine Moments for Now","event_instances":[{"all_day":false,"dt":"2018-10-31 10:00:00 -0400","end_dt":"2018-10-31 17:00:00 -0400","location_text":"The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"},{"all_day":false,"dt":"2018-11-01 10:00:00 -0400","end_dt":"2018-11-01 17:00:00 -0400","location_text":"The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"},{"all_day":false,"dt":"2018-11-02 10:00:00 -0400","end_dt":"2018-11-02 17:00:00 -0400","location_text":"The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"},{"all_day":false,"dt":"2018-11-03 10:00:00 -0400","end_dt":"2018-11-03 17:00:00 -0400","location_text":"The Ethelbert Cooper Gallery of African and African American Art\n102 Mount Auburn St.\nCambridge"}]},"doc_as_upsert":true}
```

This should work recursively for any level of nested data.

**Comment by sckott** [comment]
Maybe a reproducible example? 

Did you actually test this code?  if you install the package on your branch, then use it, it can't find `unbox()` from `jsonlite` - 

make sure to test it pleae

**Comment by sckott** [comment]
any thoughts @dpmccabe ?

**Comment by dpmccabe** [comment]
Sorry for going silent. I'll add some unit tests to this PR by this weekend.

**Comment by sckott** [comment]
thanks

**Comment by dpmccabe** [comment]
Unit test added (and passing).

**Comment by sckott** [comment]
thanks, having a look

**Comment by sckott** [comment]
LGTM

**Pull request merged**